docs/site/guide/layers.md

```diff
@@ -0,0 +1,154 @@
+# Layers
+
+Just as `Tensor` is our fundamental building block for accelerated parallel computation, most
+machine learning models and operations will be expressed in terms of the
+[`Layer`](https://github.com/tensorflow/swift-apis/blob/main/Sources/TensorFlow/Layer.swift)
+protocol. `Layer` defines an interface for types that take a differentiable input, process it, and
+produce a differentiable output. A `Layer` can contain state, such as trainable weights.
+
+`Layer` is a refinement of the `Module` protocol, with `Module` defining the more general case where
+the input to the type is not necessarily differentiable. Most components in a model will deal with
+differentiable inputs, but there are cases where types may need to conform to `Module` instead.
+
+If you create an operation has no trainable parameters within it, you'll want to define it in terms
+of `ParameterlessLayer` instead of `Layer`. 
+
+Models themselves are often defined as `Layer`s, and are regularly composed of other `Layer`s. A 
+model or subunit that has been defined as a `Layer` can be treated just like any other `Layer`, 
+allowing for the construction of arbitarily complex models from other models or subunits.
+
+To define a custom `Layer` for a model or operation of your own, you generally will follow a
+template similar to this:
+
+```swift
+public struct MyModel: Layer {
+  // Define your layers or other properties here.
+
+  // A custom initializer may be desired to configure the model.
+  public init() {}
+
+  @differentiable
+  public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+    // Define the sequence of operations performed on model input to arrive at the output.
+    return ...
+  }
+}
+```
+
+The properties of the `Layer` can host trainable components, such as weights and biases, or other
```

Comment:
I wonder if there's a better wording for "the properties... can host trainable components", which is not quite clear.

Maybe you want to say something like "Trainable components of `Layer`s, such as weights and biases or other `Layer`s, can be declared as properties".

Reply:
That sounds better to me, so I've stolen that wording.